File: app/src/main/java/com/arturogutierrez/openticator/domain/account/add/AddAccountFromCameraPresenter.java
package com.arturogutierrez.openticator.domain.account.add;

import android.content.Intent;
import com.arturogutierrez.openticator.domain.account.AccountDecoder;
import com.arturogutierrez.openticator.domain.account.camera.zxing.Intents;
import com.arturogutierrez.openticator.domain.account.interactor.AddAccountInteractor;
import com.arturogutierrez.openticator.domain.account.model.Account;
import com.arturogutierrez.openticator.interactor.DefaultSubscriber;
import com.arturogutierrez.openticator.view.presenter.Presenter;
import javax.inject.Inject;

public class AddAccountFromCameraPresenter extends DefaultSubscriber<Account> implements Presenter {

  private final AddAccountInteractor addAccountInteractorInteractor;
  private final AccountDecoder accountDecoder;
  private AddAccountFromCameraView view;

  @Inject
  public AddAccountFromCameraPresenter(AddAccountInteractor addAccountInteractorInteractor,
      AccountDecoder accountDecoder) {
    this.addAccountInteractorInteractor = addAccountInteractorInteractor;
    this.accountDecoder = accountDecoder;
  }

  public void setView(AddAccountFromCameraView view) {
    this.view = view;
  }

  @Override
  public void resume() {

  }

  @Override
  public void pause() {

  }

  @Override
  public void destroy() {
    addAccountInteractorInteractor.unsubscribe();
  }

  @Override
  public void onNext(Account account) {
    view.dismissScreen();
  }

  @Override
  public void onError(Throwable e) {
    showQRError();
  }

  public void onScanQR() {
    view.showLoading();
    view.openCaptureCode();
  }

  public void onScannedQR(Intent data) {
    String accountUri = data.getStringExtra(Intents.Scan.RESULT);
    if (accountUri == null) {
      showQRError();
      return;
    }

    decodeAccount(accountUri);
  }

  private void decodeAccount(String accountUri) {
    Account account = accountDecoder.decode(accountUri);
    if (account == null) {
      showQRError();
    } else {
      addAccountInteractorInteractor.configure(account.getName(), account.getSecret());
      addAccountInteractorInteractor.execute(this);
    }
  }

  private void showQRError() {
    view.showQRError();
    view.hideLoading();
  }
}

File: app/src/main/kotlin/com/arturogutierrez/openticator/domain/account/add/AddAccountFromCameraPresenter.kt
package com.arturogutierrez.openticator.domain.account.add

import com.arturogutierrez.openticator.domain.account.AccountDecoder
import com.arturogutierrez.openticator.domain.account.interactor.AddAccountInteractor
import com.arturogutierrez.openticator.domain.account.interactor.AddAccountInteractor.Params
import com.arturogutierrez.openticator.domain.account.model.Account
import com.arturogutierrez.openticator.interactor.DefaultSingleObserver
import com.arturogutierrez.openticator.view.presenter.Presenter
import javax.inject.Inject

class AddAccountFromCameraPresenter @Inject constructor(private val addAccountInteractorInteractor: AddAccountInteractor,
                                                        private val accountDecoder: AccountDecoder) : Presenter<AddAccountFromCameraView>() {

  override fun destroy() {
    addAccountInteractorInteractor.dispose()
  }

  fun onScanQR() {
    view?.showLoading()
    view?.openCaptureCode()
  }

  fun onScannedQR(contents: String?) {
    if (contents == null) {
      view?.hideLoading()
      return
    }

    if (contents.isEmpty()) {
      showErrorAddingAccount()
      return
    }

    decodeAccount(contents)
  }

  private fun decodeAccount(accountUri: String) {
    val account = accountDecoder.decode(accountUri)
    if (account == null) {
      showErrorAddingAccount()
    } else {
      val params = Params(account)
      addAccountInteractorInteractor.execute(params, object : DefaultSingleObserver<Account>() {
        override fun onSuccess(t: Account) {
          onAccountAdded()
        }

        override fun onError(e: Throwable) {
          onErrorAddingAccount()
        }
      })
    }
  }

  private fun onAccountAdded() {
    view?.dismissScreen()
  }

  private fun onErrorAddingAccount() {
    showErrorAddingAccount()
  }

  private fun showErrorAddingAccount() {
    view?.showQRError()
    view?.hideLoading()
  }
}
